cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,923 @@
+from openedx.core.djangoapps.content_libraries import api as libraries_api
+from openedx.core.djangoapps.content_libraries.models import ContentLibrary
+from openedx.core.djangoapps.content_staging import api as staging_api
+from xmodule.modulestore import exceptions as modulestore_exceptions
+from xmodule.modulestore.django import modulestore
+
+from .constants import CONTENT_STAGING_PURPOSE_TEMPLATE
+from .data import CompositionLevel
+from .models import ModulestoreSource, ModulestoreMigration, ModulestoreBlockSource, ModulestoreBlockMigration
+
+
+log = get_task_logger(__name__)
+
+
+class MigrationStep(Enum):
+    """
+    Strings representation the state of an in-progress modulestore-to-learning-core import.
+
+    We use these values to set UserTaskStatus.state.
+    The other possible UserTaskStatus.state values are the built-in ones:
+    UserTaskStatus.{PENDING,FAILED,CANCELED,SUCCEEDED}.
+    """
+    VALIDATING_INPUT = 'Validating migration parameters'
+    CANCELLING_OLD = 'Cancelling any redundant migration tasks'
+    LOADING = 'Loading legacy content from ModulesStore'
+    STAGING = 'Staging legacy content for import'
+    PARSING = 'Parsing staged OLX'
+    IMPORTING_ASSETS = 'Importing staged files and resources'
+    IMPORTING_STRUCTURE = 'Importing staged content structure'
+    UNSTAGING = 'Cleaning staged content'
+    MAPPING_OLD_TO_NEW = 'Saving map of legacy content to migrated content'
+    FORWARDING = 'Forwarding legacy content to migrated content'
+    POPULATING_COLLECTION = 'Assigning imported items to the specified collection'
+
+
+class _MigrationTask(UserTask):
+    """
+    Base class for migrate_to_modulestore
+    """
+
+    @staticmethod
+    def calculate_total_steps(arguments_dict):
+        """
+        Get number of in-progress steps in importing process, as shown in the UI.
+        """
+        return len(list(MigrationStep))
+
+
+@dataclass(frozen=True)
+class MigrationContext:
+    existing_source_to_target_keys: dict[UsageKey, PublishableEntity]
+    target_package_id: int
+    target_library_key: LibraryLocatorV2
+    source_context_key: CourseKey  # Note: This includes legacy LibraryLocators, which are sneakily CourseKeys.
+    composition_level: CompositionLevel
+    replace_existing: bool
+    preserve_url_slugs: bool
+    created_by: int
+    created_at: datetime
+
+    def is_already_migrated(self, source_key: UsageKey) -> bool:
+        return source_key in self.existing_source_to_target_keys
+
+    def get_existing_target(self, source_key: UsageKey) -> PublishableEntity | None:
+        return self.existing_source_to_target_keys.get(source_key)
+
+    def add_migration(self, source_key: UsageKey, target: PublishableEntity) -> None:
+        """Update the context with a new migration (keeps it current)"""
+        self.existing_source_to_target_keys[source_key] = target
+
+    def get_migrated_entities(self, base_key: str) -> set[UsageKey]:
```

Comment:
```suggestion
    def get_existing_target_entity_keys(self, base_key: str) -> set[str]:
```
We're returning entity keys (which are strings), so let's update the name and type signature to match that.